src/Featurizers/Components/WindowFeaturizerBase.h

```diff
@@ -0,0 +1,244 @@
+    ///                 it will instead wrap around back to the beginning.
+    ///
+    template <class T>
+    class CircularIterator
+        : public std::iterator<std::forward_iterator_tag,
+                               T,
+                               std::ptrdiff_t,
+                               T*,
+                               T&> {
+
+        // Assert that this class handles all the iterator traits correctly. It inherits from std::iterator so it gets these by default,
+        // but this check is here just to guarantee it is correct.
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::difference_type, std::ptrdiff_t>::value, "Make sure standard iterator handles iterator traits correctly!");
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::value_type, T>::value, "Make sure standard iterator handles iterator traits correctly!");
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::pointer, T*>::value, "Make sure standard iterator handles iterator traits correctly!");
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::reference, T&>::value, "Make sure standard iterator handles iterator traits correctly!");
+        static_assert(std::is_same<typename std::iterator_traits<CircularIterator>::iterator_category, std::forward_iterator_tag>::value, "Make sure standard iterator handles iterator traits correctly!");
+
+        // ----------------------------------------------------------------------
+        // |  Private Data
+
+        // Pointer to the first element of the underlying data structure.
+        T*                                      _itr;
+
+        // The number of elements that the underlying data structure has.
+        size_t                                  _size;
+
+        // The index this iterator is pointer it in the underlying data structure.
+        size_t                                  _cur_index;
+
+        // The maximum number of times this iterator can be incremented. Since this iterator will
+        // wrap at the end of the data structure, there is no real "end" of the data so this is
+        // used to determine when we have met the "end".
+        size_t                                  _max_increments;
+
+        // How many times we have incremented this iterator.
+        size_t                                  _cur_increment;
+
+    public:
+        // ----------------------------------------------------------------------
+        // |  Public Methods
+
+        // Default constructor, already at the "end".
+        CircularIterator()
+            : _itr(nullptr), _size(0), _cur_index(0),
+            _max_increments(0), _cur_increment(0) {
+        }
+
+        CircularIterator(T* T, size_t container_max_size, size_t max_increments, size_t starting_offset = 0)
+            : _itr(T), _size(container_max_size), _cur_index(starting_offset),
```

Comment:
>_itr [](start = 14, length = 4)

Raise exceptions if t is nullptr, container_max_size == 0, max_increments == 0, or starting_offset > container_max_size. Also raise error if max_increments > container_max_size. Add corresponding tests.

Reply:
we use max_increments = 0 with our "end" iterator, so we can't raise an exception on that, but I will add in the others.

Reply:
Also, is there a problem if max_increments is greater than max_size? That would allow them to loop over the data multiple times if desired. Or do you think thats bad behavior so we just shoudln't allow it?

Reply:
I'd recommend adding comments if this is a scenario that you are supporting.

---
In reply to: [385345847](https://github.com/microsoft/FeaturizersLibrary/pull/154#discussion_r385345847) [](ancestors = 385345847)

Reply:
Added an exception to make sure `max_increments` is not greater than `max_size`